provisioning/baremetal_config.go

```diff
@@ -0,0 +1,84 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provisioning
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/golang/glog"
+	metal3iov1alpha1 "github.com/openshift/cluster-baremetal-operator/api/v1alpha1"
+)
+
+const (
+	// BaremetalProvisioningCR is the name of the provisioning config resource.
+	BaremetalProvisioningCR = "provisioning-configuration"
+)
+
+// ValidateBaremetalProvisioningConfig validates the contents of the baremetal Provisioning CR.
+func ValidateBaremetalProvisioningConfig(prov *metal3iov1alpha1.Provisioning) error {
+	provisioningNetworkMode := getProvisioningNetworkMode(prov)
+	switch provisioningNetworkMode {
+	case metal3iov1alpha1.ProvisioningNetwork("Managed"):
+		glog.V(1).Info("Provisioning Network Managed")
+		allowedAbsent := []string{"ProvisioningNetwork", "ProvisioningDHCPExternal"}
+		return validateConfig(prov, allowedAbsent)
+	case metal3iov1alpha1.ProvisioningNetwork("Unmanaged"):
+		glog.V(1).Info("Provisioning Network Unmanaged")
+		allowedAbsent := []string{"ProvisioningNetwork", "ProvisioningDHCPExternal", "ProvisioningDHCPRange"}
+		return validateConfig(prov, allowedAbsent)
+	case metal3iov1alpha1.ProvisioningNetwork("Disabled"):
+		glog.V(1).Info("Provisioning Network Disabled")
+		allowedAbsent := []string{"ProvisioningInterface", "ProvisioningDHCPExternal", "ProvisioningDHCPRange"}
+		return validateConfig(prov, allowedAbsent)
+	}
+	return nil
+}
+
+func getProvisioningNetworkMode(prov *metal3iov1alpha1.Provisioning) metal3iov1alpha1.ProvisioningNetwork {
+	provisioningNetworkMode := prov.Spec.ProvisioningNetwork
+	if provisioningNetworkMode == "" {
+		// Set it to the default Managed mode
+		provisioningNetworkMode = metal3iov1alpha1.ProvisioningNetwork("Managed")
+		if prov.Spec.ProvisioningDHCPExternal {
+			glog.V(1).Infof("ProvisioningDHCPExternal is being deprecated in favor of ProvisioningNetwork and will be removed in the next release")
+			provisioningNetworkMode = metal3iov1alpha1.ProvisioningNetwork("Unmanaged")
+		}
+		glog.V(1).Infof("ProvisioningNetwork config not provided. Based on ProvisioningDHCPExternal setting it to: %s", provisioningNetworkMode)
+	}
+	return provisioningNetworkMode
+}
+
+func isAllowedTobeMissing(slice []string, value string) bool {
+	for _, item := range slice {
+		if item == value {
+			return true
+		}
+	}
+	return false
+}
+
+func validateConfig(prov *metal3iov1alpha1.Provisioning, optionalConfig []string) error {
+	specConfig := reflect.ValueOf(prov.Spec)
```

Comment:
Using reflection here feels like an unnecessary obfuscation of what the function is doing. The earlier structure with inline references to the fields was fine, it just needed to be restructured to allow it to report the name of the missing value. That could be done clearly using a slice of a struct with a field name and a field value.

```

for _, toTest := range []struct{
    Name string
    Value string
  }{
    {Name: "ProvisioningInterface", Value: prov.Spec.ProvisioningInterface},
  }{
  if toTest.Value == "" {
    return fmt.Errorf("%s is required but is empty", toTest.Name)
  }
}
```

Reply:
The reason I decided to use reflect are the following: 
If and when we add new fields to the Spec, it could be of any type (not just string). For example, our earlier change to make ProvisioningNetwork into a enum type instead of string might break above code (haven't actually tested it). Using reflect would allow us to introspect our own data structure and do the right thing based on the type of the field in the struct.
Also, when the Provisioning API changes (and specifically the Spec field), all we will have to do is change the allowedAbsent slice per mode.

If you still feel that this is unnecessary, I am happy to use something similar to what is suggested.

Reply:
If we end up having to compare different types, I think I would rather just use separate conditional statements outside of a loop. Kiran pointed out on an earlier review that there are additional checks we need to do besides ensuring that values are present, and when we start adding those it doesn't seem likely that using reflection would be necessary or helpful.

Let's try to keep the implementation as simple as possible, even if it means a little repetition.

Reply:
Made it simpler by having a separate validation function for each ProvisioningNetwork mode.